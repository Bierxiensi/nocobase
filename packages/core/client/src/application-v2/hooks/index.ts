export * from './useApp';
export * from './useAppPluginLoad';
export * from './useRouter';
