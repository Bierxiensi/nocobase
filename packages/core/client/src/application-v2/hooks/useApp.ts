import { useContext } from 'react';
import { Application } from '../Application';
import { ApplicationContext } from '../context';

export const useApp = () => {
  return useContext(ApplicationContext) || ({} as Application);
};
