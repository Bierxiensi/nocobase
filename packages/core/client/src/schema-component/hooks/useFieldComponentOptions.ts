import { useFieldSchema } from '@formily/react';
import { useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import { useCollection, useCollectionManager } from '../../collection-manager';

export const useFieldComponentOptions = () => {
  const { getCollectionJoinField, getCollection } = useCollectionManager();
  const fieldSchema = useFieldSchema();
  const { getField } = useCollection();
  const collectionField = getField(fieldSchema['name']) || getCollectionJoinField(fieldSchema['x-collection-field']);
  const { t } = useTranslation();
  const { label } = fieldSchema['x-component-props']?.fieldNames || {};

  const fieldComponentOptions = useMemo(() => {
    if (!collectionField || !collectionField?.interface) {
      return;
    }

    if (!['o2o', 'oho', 'obo', 'o2m', 'linkTo', 'm2o', 'm2m'].includes(collectionField.interface)) return;

    const collection = getCollection(collectionField.target);
    if (collection?.template === 'file') {
      return [
        { label: t('Record picker'), value: 'CollectionField' },
        { label: t('Select'), value: 'AssociationSelect' },
      ];
    }

    switch (collectionField.interface) {
      case 'o2m':
        return [
          { label: t('Record picker'), value: 'RecordPicker' },
          { label: t('Subtable'), value: 'SubTable' },
          { label: t('Select'), value: 'AssociationSelect' },
          { label: t('Subform'), value: 'SubForm' },
        ];

      case 'm2o':
      case 'm2m':
      case 'linkTo':
        return [
          { label: t('Record picker'), value: 'RecordPicker' },
          { label: t('Select'), value: 'AssociationSelect' },
          { label: t('Subform'), value: 'SubForm' },
        ];

      default:
        return [
          { label: t('Record picker'), value: 'RecordPicker' },
          { label: t('Select'), value: 'AssociationSelect' },
          { label: t('Subform'), value: 'SubForm' },
        ];
    }
  }, [t, collectionField?.interface, label]);

  return fieldComponentOptions;
};
