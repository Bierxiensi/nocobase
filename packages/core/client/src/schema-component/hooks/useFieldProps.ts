import { useContext, useRef } from 'react';
import type { Schema} from '@formily/react';
import { SchemaExpressionScopeContext, SchemaOptionsContext } from '@formily/react';

export const useFieldProps = (schema: Schema) => {
  const options = useContext(SchemaOptionsContext);
  const scope = useContext(SchemaExpressionScopeContext);
  const scopeRef = useRef<any>();
  scopeRef.current = scope;
  return schema.toFieldProps({
    ...options,
    get scope() {
      return {
        ...options.scope,
        ...scopeRef.current,
      };
    },
  }) as any;
};
