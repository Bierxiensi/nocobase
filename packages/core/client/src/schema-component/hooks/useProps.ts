interface Options {
  arrayMerge?(target: any[], source: any[], options?: Options): any[];
  clone?: boolean;
  assign?: boolean;
  customMerge?: (key: string, options?: Options) => ((x: any, y: any) => any) | undefined;
  isMergeableObject?(value: object): boolean;
  cloneUnlessOtherwiseSpecified?: (value: any, options: Options) => any;
}

const useDef = () => ({});
export const useProps = (originalProps: any = {}) => {
  const { useProps: useDynamicProps = useDef, ...others } = originalProps;
  const dynamicProps = useDynamicProps();
  return { ...dynamicProps, ...others };
};
