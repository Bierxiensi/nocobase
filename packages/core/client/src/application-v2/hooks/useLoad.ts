import { useEffect, useState } from 'react';
import { useApp } from './useApp';

export const useLoad = () => {
  const app = useApp();
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    setLoading(true);
    (async () => {
      await app.load();
      setLoading(false);
    })();
  }, []);
  return loading;
};
