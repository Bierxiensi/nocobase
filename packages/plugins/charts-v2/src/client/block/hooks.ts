import { useContext } from 'react';
import { ChartConfigContext } from './ChartConfigure';
import { useCollectionManager } from '@nocobase/client';
import { ISchema, Schema } from '@formily/react';
import { useTranslation } from 'react-i18next';
import { operators } from '@nocobase/client';

export type FieldOption = {
  value: string;
  label: string;
  key: string;
  alias?: string;
  name?: string;
  type?: string;
  interface?: string;
  uiSchema?: ISchema;
};

export const useFields = () => {
  const { t } = useTranslation();
  const { current } = useContext(ChartConfigContext);
  const { collection } = current || {};
  const { getCollectionFields } = useCollectionManager();
  const fields = (getCollectionFields(collection) || [])
    .filter((field) => {
      return !['set', 'password', 'json', 'belongsTo', 'hasMany', 'belongsToMany', 'hasOne'].includes(field.type);
    })
    .map((field) => ({
      key: field.key,
      label: field.uiSchema?.title || field.name,
      value: field.name,
      ...field,
    }));
  return Schema.compile(fields, { t });
};

export const useFilterOptions = (fields: FieldOption[]) => {
  const { getInterface } = useCollectionManager();
  const interfaceMap = {
    createdAt: 'createdAt',
    updatedAt: 'updatedAt',
    createdById: 'createdBy',
    updatedById: 'updatedBy',
  };

  const options = [];
  fields.forEach((field) => {
    let ops = [];
    const fieldInterface = getInterface(field.interface || interfaceMap[field.name]);
    if (fieldInterface) {
      ops = fieldInterface?.filterable?.operators || [];
    } else {
      ops = operators[field.type] || [];
    }
    if (!ops.length) {
      return;
    }
    options.push({
      name: field.value,
      title: field.label,
      schema: field.uiSchema,
      operators: ops,
    });
  });
  return options;
};
