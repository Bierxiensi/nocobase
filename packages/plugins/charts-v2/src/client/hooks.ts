import { useContext } from 'react';
import { ChartConfigContext, SelectedField } from './block/ChartConfigure';
import { useCollectionManager } from '@nocobase/client';
import { ISchema, Schema } from '@formily/react';
import { useTranslation } from 'react-i18next';
import { operators } from '@nocobase/client';
import formatters from './block/formatters';
import transformers from './block/transformers';
import { lang } from './locale';
import { ChartRendererProps, QueryProps } from './renderer';

export type FieldOption = {
  value: string;
  label: string;
  key: string;
  alias?: string;
  name?: string;
  type?: string;
  interface?: string;
  uiSchema?: ISchema;
};

export const useFields = (collection?: string) => {
  const { t } = useTranslation();
  const { current } = useContext(ChartConfigContext);
  if (!collection) {
    collection = current?.collection || {};
  }
  const { getCollectionFields } = useCollectionManager();
  const fields = (getCollectionFields(collection) || [])
    .filter((field) => {
      return !['belongsTo', 'hasMany', 'belongsToMany', 'hasOne'].includes(field.type) && field.interface;
    })
    .map((field) => ({
      key: field.key,
      label: field.uiSchema?.title || field.name,
      value: field.name,
      ...field,
    }));
  return Schema.compile(fields, { t }) as FieldOption[];
};

export const getAllFields = (fields: FieldOption[], field: any) => {
  // When field alias is set, appends it to the field list
  const appendAlias = (selectedFields: SelectedField[]) => {
    return selectedFields
      .filter((selectedField) => selectedField.alias)
      .map((selectedField) => {
        const fieldProps = fields.find((field) => field.name === selectedField.field);
        return {
          ...fieldProps,
          key: selectedField.alias,
          label: selectedField.alias,
          value: selectedField.alias,
        };
      });
  };
  const query = field.query('query').get('value') || {};
  const measures = query.measures || [];
  const dimensions = query.dimensions || [];
  const aliasFields = [...appendAlias(measures), ...appendAlias(dimensions)];
  // unique
  const map = new Map([...fields, ...aliasFields].map((item) => [item.value, item]));
  const allFields = [...map.values()];
  return allFields;
};

export const useChartFields = (fields: FieldOption[]) => (field: any) => {
  const allFields = getAllFields(fields, field);
  /**
   * chartFields is used for configuring chart fields
   * since the default alias is field display name, we need to set the option values to field display name
   * see also: 'appendAliasToQuery' function in 'renderer/ChartRenderer.tsx'
   */
  const chartFields = allFields.map((field) => ({
    ...field,
    value: field.label,
  }));
  field.dataSource = chartFields;
};

export const useFormatters = (fields: FieldOption[]) => (field: any) => {
  const selectedField = field.query('.field').get('value');
  if (!selectedField) {
    field.dataSource = [];
    return;
  }
  let options = [];
  const fieldInterface = fields.find((field) => field.name === selectedField)?.interface;
  switch (fieldInterface) {
    case 'datetime':
    case 'createdAt':
    case 'updatedAt':
      options = formatters.datetime;
      break;
    case 'date':
      options = formatters.date;
      break;
    case 'time':
      options = formatters.time;
      break;
    default:
      options = [];
  }
  field.dataSource = options;
};

export const useCollectionOptions = () => {
  const { t } = useTranslation();
  const { collections } = useCollectionManager();
  const options = collections.map((collection: { name: string; title: string }) => ({
    label: collection.title,
    value: collection.name,
    key: collection.name,
  }));
  return Schema.compile(options, { t });
};

export const useFieldTypes = (fields: FieldOption[]) => (field: any) => {
  const selectedField = field.query('.field').get('value');
  const allFields = getAllFields(fields, field);
  const fieldProps = allFields.find((field) => field.label === selectedField);
  const supports = Object.keys(transformers);
  field.dataSource = supports.map((key) => ({
    label: lang(key),
    value: key,
  }));
  const map = {
    createdAt: 'datetime',
    updatedAt: 'datetime',
    double: 'number',
    integer: 'number',
  };
  const fieldInterface = fieldProps?.interface;
  const fieldType = fieldProps?.type;
  const key = map[fieldInterface] || map[fieldType] || fieldType;
  if (supports.includes(key)) {
    field.setState({
      value: key,
      disabled: true,
    });
    return;
  }
  field.setState({
    value: null,
    disabled: false,
  });
};

export const useTransformers = (field: any) => {
  const selectedType = field.query('.type').get('value');
  if (!selectedType) {
    field.dataSource = [];
    return;
  }
  const options = Object.keys(transformers[selectedType] || {}).map((key) => ({
    label: lang(key),
    value: key,
  }));
  field.dataSource = options;
};

export const useQueryWithAlias = (fields: FieldOption[], query: QueryProps) => {
  // If alias is not set, use field title (display name instead of field name) as alias
  const appendAlias = (selectedFields: SelectedField[]) => {
    return selectedFields
      .filter((item) => item.field)
      .map((item) => {
        const field = fields.find((field) => field.name === item.field);
        return {
          ...item,
          alias: item.alias || field?.label,
        };
      });
  };
  const { dimensions = [], measures = [] } = query || {};
  return {
    ...query,
    dimensions: appendAlias(dimensions),
    measures: appendAlias(measures),
  };
};

export const useFieldTransformer = (transform: ChartRendererProps['transform']) => {
  return (transform || [])
    .filter((item) => item.field && item.type && item.format)
    .reduce((meta, item) => {
      const formatter = transformers[item.type][item.format];
      meta[item.field] = {
        formatter: (val: any) => formatter(val),
      };
      return meta;
    }, {});
};
