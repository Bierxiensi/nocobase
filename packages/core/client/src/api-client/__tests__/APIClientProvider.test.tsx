import React from 'react';
import { render } from '@testing-library/react';
import { APIClientProvider } from '../APIClientProvider';
import { isFunction, isBoolean } from 'lodash';
import { APIClient } from '../APIClient';
import { useRequest } from '../hooks/useRequest';

const ComponentA = () => {
  const { data, loading, refresh, run, params } = useRequest({
    url: 'users:get',
    method: 'get',
  });
  return <div>{`loading:${isBoolean(loading)}, refresh:${isFunction(refresh)}, run:${isFunction(run)}`}</div>;
};

describe('APIClientProvider', () => {
  it('useRequest is ok', () => {
    const apiClient = new APIClient({
      baseURL: 'http://localhost/api/',
    });
    const { container } = render(
      <APIClientProvider apiClient={apiClient}>
        <ComponentA />
      </APIClientProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
