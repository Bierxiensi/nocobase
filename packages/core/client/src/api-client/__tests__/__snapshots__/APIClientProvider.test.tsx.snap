// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APIClientProvider useRequest is ok 1`] = `
<div>
  <div>
    loading:true, refresh:true, run:true
  </div>
</div>
`;
